pkg/ingester/otlp/convert.go

```diff
@@ -6,197 +6,238 @@ import (
 		StartLine:  of.StartLine,
 	}
+	p.dst.Function = append(p.dst.Function, gf)
+	gf.Id = uint64(len(p.dst.Function))
+	p.functionMap[of] = gf.Id
+	return gf.Id
 }
 
-func convertSampleBack(p *otelProfile.Profile, os *otelProfile.Sample, locationIndexes []int64, addstr func(s string) int64) *googleProfile.Sample {
+func (p *profileBuilder) convertSampleBack(os *otelProfile.Sample) *googleProfile.Sample {
 	gs := &googleProfile.Sample{
 		Value: os.Value,
 	}
 
 	if len(gs.Value) == 0 {
 		gs.Value = []int64{int64(len(os.TimestampsUnixNano))}
 	}
-	convertSampleAttributesToLabelsBack(p, os, gs, addstr)
+	p.convertSampleAttributesToLabelsBack(os, gs)
 
 	for i := os.LocationsStartIndex; i < os.LocationsStartIndex+os.LocationsLength; i++ {
-		gs.LocationId = append(gs.LocationId, uint64(locationIndexes[i]+1))
+		gs.LocationId = append(gs.LocationId, p.convertLocationBack(p.src.Location[p.src.LocationIndices[i]]))
 	}
 
+	p.dst.Sample = append(p.dst.Sample, gs)
+
 	return gs
 }
 
-func convertSampleAttributesToLabelsBack(p *otelProfile.Profile, os *otelProfile.Sample, gs *googleProfile.Sample, addstr func(s string) int64) {
+func (p *profileBuilder) convertSampleAttributesToLabelsBack(os *otelProfile.Sample, gs *googleProfile.Sample) {
 	gs.Label = make([]*googleProfile.Label, 0, len(os.Attributes))
 	for _, attribute := range os.Attributes {
-		att := p.AttributeTable[attribute]
+		att := p.src.AttributeTable[attribute]
+		if att.Key == serviceNameKey {
+			continue
+		}
 		if att.Value.GetStringValue() != "" {
 			gs.Label = append(gs.Label, &googleProfile.Label{
-				Key: addstr(att.Key),
-				Str: addstr(att.Value.GetStringValue()),
+				Key: p.addstr(att.Key),
+				Str: p.addstr(att.Value.GetStringValue()),
 			})
 		}
 	}
 }
 
 // convertMappingsBack converts a slice of OpenTelemetry Mapping entries to Google Mapping entries.
-func convertMappingsBack(otelMappings []*otelProfile.Mapping) []*googleProfile.Mapping {
-	googleMappings := make([]*googleProfile.Mapping, len(otelMappings))
-	for i, om := range otelMappings {
-		googleMappings[i] = &googleProfile.Mapping{
-			Id:              uint64(i + 1), // Assuming direct mapping of IDs
-			MemoryStart:     om.MemoryStart,
-			MemoryLimit:     om.MemoryLimit,
-			FileOffset:      om.FileOffset,
-			Filename:        om.Filename, // Assume direct use; may need conversion if using indices
-			BuildId:         om.BuildId,
-			HasFunctions:    om.HasFunctions,
-			HasFilenames:    om.HasFilenames,
-			HasLineNumbers:  om.HasLineNumbers,
-			HasInlineFrames: om.HasInlineFrames,
-		}
-	}
-	return googleMappings
+func (p *profileBuilder) convertMappingBack(om *otelProfile.Mapping) uint64 {
+	if i, ok := p.mappingMap[om]; ok {
+		return i
+	}
+
+	gm := &googleProfile.Mapping{
+		MemoryStart:     om.MemoryStart,
+		MemoryLimit:     om.MemoryLimit,
+		FileOffset:      om.FileOffset,
+		Filename:        p.addstr(p.src.StringTable[om.Filename]),
+		BuildId:         p.addstr(p.src.StringTable[om.BuildId]),
```

Comment:
This isn't going to work with the [new spec](https://github.com/open-telemetry/opentelemetry-proto/blob/2bd940b2b77c1ab57c27166af21384906da7bb2b/opentelemetry/proto/profiles/v1development/profiles.proto#L402).

However, not related to this PR since we're relying on `om.BuildId` before. I created an issue last week https://github.com/grafana/pyroscope-squad/issues/295